File: modules/benchmarks/arrow-scala-benchmarks/src/main/scala/zio/AttemptNonRaised.scala
package arrow.benchmarks.effects.scala.zio

import scalaz.zio._

object AttemptNonRaised {

  def ioLoopHappy(size: Int, i: Int): Task[Int] =
    if (i < size) {
      IO.effect {
        i + 1
      }.either.flatMap { result =>
          result.fold[Task[Int]](IO.fail, ioLoopHappy(size, _))
        }
    } else IO.succeed(1)

  def run(size: Int) = ZIORTS.unsafeRun(ioLoopHappy(size, 0))

}

File: modules/benchmarks/arrow-benchmarks-fx/src/jmh/kotlin/arrow/benchmarks/AttemptNonRaised.kt
package arrow.benchmarks

import arrow.fx.IO
import org.openjdk.jmh.annotations.Benchmark
import org.openjdk.jmh.annotations.CompilerControl
import org.openjdk.jmh.annotations.Fork
import org.openjdk.jmh.annotations.Measurement
import org.openjdk.jmh.annotations.Param
import org.openjdk.jmh.annotations.Scope
import org.openjdk.jmh.annotations.State
import org.openjdk.jmh.annotations.Warmup
import java.util.concurrent.TimeUnit

@State(Scope.Thread)
@Fork(2)
@Warmup(iterations = 10, time = 1, timeUnit = TimeUnit.SECONDS)
@Measurement(iterations = 10)
@CompilerControl(CompilerControl.Mode.DONT_INLINE)
open class AttemptNonRaised {

  @Param("10000")
  var size: Int = 0

  private fun ioLoopHappy(size: Int, i: Int): IO<Int> =
    if (i < size) {
      IO { i + 1 }.attempt().flatMap {
        it.fold(IO.Companion::raiseError) { n -> ioLoopHappy(size, n) }
      }
    } else IO.just(1)

  @Benchmark
  fun io(): Int =
    ioLoopHappy(size, 0).unsafeRunSync()

  @Benchmark
  fun cats(): Any =
    arrow.benchmarks.effects.scala.cats.AttemptNonRaised.ioLoopHappy(size, 0).unsafeRunSync()

  @Benchmark
  fun zio(): Any =
    arrow.benchmarks.effects.scala.zio.AttemptNonRaised.run(size)
}
